### PR TITLE
[Extend] version-automation to `rockcraft.yaml`

### PR DESCRIPTION
**Pull request by rudra-iitm**
Extended `desktop-snap` to support `version automation` in `rockraft.yaml` also.

- Users need to define their version schema in `rock-version-schema`, and all other rules will follow as in `snap`, such as:
    - The version should be defined in `metadata`, and the `leading upstream component` should be defined in the `adopt-info` section.
- The automated version format is the same as that of snap `<upstream-version>-<package-release>`.
- When this workflow runs, it checks if there are any updates in the source tag of any part and then updates the version accordingly:
    - If there is a change in the source tag of the leading upstream component, it sets the package release number to 1 and the upstream version to match the leading upstream component.
    - If there is any change in any other source tag, it increments the package release number by 1.

### CODE REVIEW

**Comment by rudra-iitm** [comment]
@sergio-costas, pls review this PR also

**Comment by rudra-iitm** [comment]
@sergio-costas, any update on this PR?

**Comment by sergio-costas** [comment]
Sorry, didn't receive the notification. I'll check it now.

File: updatesnap/SnapVersionModule/snap_version_module.py [hunk 1]
```diff
@@ -62,6 +62,32 @@ def process_snap_version_data(upstreamversion, snap_name, version_schema, has_up
     return f"{upstreamversion}-{packagerelease}"
 
 
+def process_rock_version_data(upstreamversion, prevversion, version_schema, has_update):
```
Comment (sergio-costas) [inline updatesnap/SnapVersionModule/snap_version_module.py]: `upstream_version` and `previous_version`

File: updatesnap/SnapVersionModule/snap_version_module.py [hunk 2]
```diff
@@ -62,6 +62,32 @@ def process_snap_version_data(upstreamversion, snap_name, version_schema, has_up
...
+    """ Returns processed rock version"""
+
+    match = re.match(version_schema, upstreamversion)
+    if not match:
+        logging.warning("Version schema does not match with rock repository version")
+        return None
+    upstreamversion = match.group(1).replace('_', '.')
+
+    def version_tuple(v):
```
Comment (sergio-costas) [inline updatesnap/SnapVersionModule/snap_version_module.py]: Do we really need to create a function just for this?

**Review by sergio-costas — CHANGES_REQUESTED** [review]

File: updatesnap/SnapVersionModule/snap_version_module.py (hunk 2 above)
Comment (rudra-iitm) [reply updatesnap/SnapVersionModule/snap_version_module.py]: Removed function and now using inline tuple to resolve version

File: updatesnap/SnapVersionModule/snap_version_module.py (hunk 1 above)
Comment (rudra-iitm) [reply updatesnap/SnapVersionModule/snap_version_module.py]: Updated

File: updatesnap/SnapVersionModule/snap_version_module.py [hunk 3]
```diff
@@ -62,6 +62,33 @@ def process_snap_version_data(upstreamversion, snap_name, version_schema, has_up
...
+    match = re.match(version_schema, upstream_version)
+    if not match:
+        logging.warning("Version schema does not match with rock repository version")
+        return None
+    upstream_version = match.group(1).replace('_', '.')
+
+    def version_tuple(v):
+        return tuple(map(int, v.split('.')))
+
+    upstream_tuple = version_tuple(upstream_version)
```
Comment (sergio-costas) [inline updatesnap/SnapVersionModule/snap_version_module.py]: Ehm... Is this a problem in github, or the function is really still there?

**Review by sergio-costas — CHANGES_REQUESTED** [review]

File: updatesnap/SnapVersionModule/snap_version_module.py (hunk 3 above)
Comment (rudra-iitm) [reply updatesnap/SnapVersionModule/snap_version_module.py]: Sorry I forgot to remove that functions... Now fixed it

**Review by sergio-costas — APPROVED** [review]
Looks fine.

**Comment by sergio-costas** [comment]
Thanks for the changes.

**Pull request merged**